compiler/src/dotty/tools/dotc/ast/Desugar.scala

```diff
@@ -1114,8 +1117,8 @@ object desugar {
         Apply(Select(Apply(Ident(nme.StringContext), strs), id), elems)
       case InfixOp(l, op, r) =>
         if (ctx.mode is Mode.Type)
-          if (!op.isBackquoted && op.name == tpnme.raw.AMP) AndTypeTree(l, r)     // l & r
-          else if (!op.isBackquoted && op.name == tpnme.raw.BAR) OrTypeTree(l, r) // l | r
+          if (!op.isBackquoted && op.name == tpnme.raw.AMP) AndTypeTree(checkWildcard(l), checkWildcard(r))     // l & r
```

Comment:
I wonder if we shouldn't do this in Parser directly. This desugaring is kind of weird anyway, since it's not really about sugar but about how to parse something: we could just emit AndTypeTree or OrTypeTree as appropriate directly in Parser

Reply:
The reason to do it this way is that `op` is an identifier with a position. Converted to And/OrTypeTree, that information is lost.

Reply:
👍, trying this out.

It took me a while to give up searching `OrTypeTree` in the `Parser`, and honestly I was very confused that wasn't already the case.

Reply:
> The reason to do it this way is that op is an identifier with a position. Converted to And/OrTypeTree, that information is lost.

But is that information useful for anything?

Reply:
FWIW, changing this passes CI, see #4424.

Reply:
> But is that information useful for anything?

Go to definition in the IDE?